theano/tensor/basic.py

```diff
@@ -4372,7 +4372,7 @@ def infer_shape(self, node, ishapes):
                     if ele == -1:
                         requ[i] = missing
             elif crit == 1:  # we reshape to -1
-                requ = [mul(*ishapes[0])]
+                requ = [mul(*ishapes[0])] if ishapes[0] else [1]
```

Comment:
>I was talking about changing mul() interface do not crash when it get no inputs.

Oh ok. I agree with you not to change mul's interface. It should fail when there is no argument, just as python's ```sum``` does for instance.

> Can you tell more on what in the Reshape.infer_shape is broken? If it is broken, maybe we should wait for it to be fixed before converting Flatten to Reshape.

It's using Shape_i() which makes TestInferShape fail as there is still some Reshapes in shape's graph after optimization. See issue #3420 

> I don't understand your questions. For scalar (ndim=0), We simply have an empty tuple.

Yep sure. I think I understand now why you say the else cause isn't good. It's because it should be ```[mul(*ishapes[0])] if ishapes[0] else []```, isn't? Otherwise we are turning a scalar into a 1 dimensional vector.